src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs

```diff
@@ -0,0 +1,536 @@
+
+    // Always soft-select these completion items.  Also, never filter down.
+    private static readonly CompletionItemRules s_rules = CompletionItemRules.Create(
+        selectionBehavior: CompletionItemSelectionBehavior.SoftSelection,
+        filterCharacterRules: ImmutableArray.Create(CharacterSetModificationRule.Create(CharacterSetModificationKind.Replace, Array.Empty<char>())));
+
+    // The space between type and parameter name.
+    // void TestMethod(int // <- space after type name
+    public ImmutableHashSet<char> TriggerCharacters { get; } = ImmutableHashSet.Create(' ');
+
+    public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
+    {
+        if (trigger.Kind == CompletionTriggerKind.Insertion)
+        {
+            return TriggerCharacters.Contains(trigger.Character);
+        }
+
+        return false;
+    }
+
+    public override Task<CompletionDescription> GetDescriptionAsync(Document document, CompletionItem item, CancellationToken cancellationToken)
+    {
+        if (!item.Properties.TryGetValue(DescriptionKey, out var description))
+        {
+            return Task.FromResult<CompletionDescription>(null);
+        }
+
+        return Task.FromResult(CompletionDescription.Create(
+            ImmutableArray.Create(new TaggedText(TextTags.Text, description))));
+    }
+
+    public override Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey, CancellationToken cancellationToken)
+    {
+        // These values have always been added by us.
+        var startString = item.Properties[StartKey];
+        var lengthString = item.Properties[LengthKey];
+        var newText = item.Properties[NewTextKey];
+
+        // This value is optionally added in some cases and may not always be there.
+        item.Properties.TryGetValue(NewPositionKey, out var newPositionString);
+
+        return Task.FromResult(CompletionChange.Create(
+            new TextChange(new TextSpan(int.Parse(startString, CultureInfo.InvariantCulture), int.Parse(lengthString, CultureInfo.InvariantCulture)), newText),
+            newPositionString == null ? null : int.Parse(newPositionString, CultureInfo.InvariantCulture)));
+    }
+
+    public override async Task ProvideCompletionsAsync(CompletionContext context)
+    {
+        if (context.Trigger.Kind is not CompletionTriggerKind.Insertion)
+        {
+            return;
+        }
+
+        var position = context.Position;
+
+        var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+        if (root == null)
+        {
+            return;
+        }
+        var token = root.FindToken(position);
+
+        // If space is before a close paren then change current token to the previous argument.
+        if (token.IsKind(SyntaxKind.CloseParenToken) || token.IsKind(SyntaxKind.CommaToken))
```

Comment:
this si not validating that the space is actually before thsee guys.  for example `) $$ `.  FindToken here would find the close paren, but you would be after it.